### PR TITLE
webserver: refactor, unification of handlers and bug fixes

### PR DESCRIPTION
**Pull request by Montellese**
While working on some webserver related stuff I realized that all of the special handling I added for HTTP features like `Cache-Control`, `If-Modified-Since`, `If-Unmodified-Since`, `If-Range` and the whole range support is limited to the case where the client requests an existing file and that the logic didn't apply to "generated" content.
For that reason I have refactored the webserver implementation and all the HTTP request handlers we currently have (VFS, JSON-RPC, web interfaces, ...) to use the same logic to handle the above mentioned HTTP features where it makes sense.

Because all these features are rather hard to test and make sure that they (still) work I decided to write unit tests to cover the usual and edge cases of these features. While writing these unit tests I came across a few cases that aren't properly handled by the current implementation. The unit tests use our `CCurlFile` implementation to send HTTP requests to our own webserver with test specific headers and then checks if the result is as expected. For that to work I had to make a few adjusments to the unit test setup.

After the refactor all 26 unit tests (especially HTTP range support) pass (with the old code only 19 passed). I've also written new helper classes to support HTTP range support (header parsing, data generation etc.) for which I've also written 30 unit tests to hopefully cover most possible use cases. An additional features supported by the new implementation is that it merges multiple overlapping or back-to-back ranges and can handle ranges defined in an arbitrary order.

EDIT: This requires Xcode changes.

### CODE REVIEW

**Comment by Montellese** [comment]
Added some documentation to the newly added HttpRangeUtils.

**Comment by topfs2** [comment]
I haven't reviewed the code but +1 on refactoring it. Sounds like its a
nice patch!
Den 14 jan 2015 08:10 skrev "Sascha Montellese" <notifications@github.com>:

File: xbmc/network/httprequesthandler/HTTPFileHandler.cpp [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+ */
+
+#include "system.h"
+#include "HTTPFileHandler.h"
+#include "filesystem/File.h"
+#include "utils/Mime.h"
+#include "utils/StringUtils.h"
+#include "utils/URIUtils.h"
+
+using namespace std;
```
Comment (Paxxi) [inline xbmc/network/httprequesthandler/HTTPFileHandler.cpp]: This isn't needed, you already prefixed all std functions

File: xbmc/network/httprequesthandler/HTTPFileHandler.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "IHTTPRequestHandler.h"
```
Comment (Paxxi) [inline xbmc/network/httprequesthandler/HTTPFileHandler.h]: should include <string>

**Comment by Paxxi** [comment]
Nice work! Especially the huge amount of tests you added. I have some nitpicking.
1. the use of std:: is inconsistent through single files
2. I'd like to see some comments/warnings around these two response types HTTPMemoryDownloadFreeNoCopy, HTTPMemoryDownloadFreeCopy that the data needs to be malloced and not newed to avoid data leaks.

Again, nice work!

**Comment by Montellese** [comment]
@Paxxi: Thanks for the review. Yeah std namespace is a mess because I often used `using namespace std;` in my old code and left it that way in code I copied/moved but in new code I usually prefer writing `std::string` or `std::map` in the CPP file as well to be consistent with the header file. I can add another cosmetic commit at the end which cleans it up in the webserver related code.
And I can add the requested commets. That enum is basically just a wrapper for the different methods provided by libmicrohttpd.

**Comment by Montellese** [comment]
* Made usage of `std::` namespace cleaner in the `IHTTPRequestHandler` interface implementations.
* Added documentation for the `HTTPResponseType` enums as requested
* Added an additional commit that takes care of an additional potential error case

**Comment by Montellese** [comment]
jenkins build this please

**Comment by Montellese** [comment]
Obviously the build fails on OSX/IOS/ATV2 (forgot that it needs Xcode changes)...
@Memphiz: could you add the new files to the Xcode project(s)?

**Comment by Montellese** [comment]
@Memphiz: Thanks for the quick Xcode update.

jenkins build this please
